crates/bevy_animation/src/lib.rs

```diff
@@ -192,100 +243,161 @@ pub fn animation_player(
-                        // PERF: finding the current keyframe can be optimised
-                        let step_start = match curve
-                            .keyframe_timestamps
-                            .binary_search_by(|probe| probe.partial_cmp(&elapsed).unwrap())
-                        {
-                            Ok(n) if n >= curve.keyframe_timestamps.len() - 1 => continue, // this curve is finished
-                            Ok(i) => i,
-                            Err(0) => continue, // this curve isn't started yet
-                            Err(n) if n > curve.keyframe_timestamps.len() - 1 => continue, // this curve is finished
-                            Err(i) => i - 1,
-                        };
-                        let ts_start = curve.keyframe_timestamps[step_start];
-                        let ts_end = curve.keyframe_timestamps[step_start + 1];
-                        let lerp = (elapsed - ts_start) / (ts_end - ts_start);
-
-                        // Apply the keyframe
+fn _run_animation_player(
+    entity: Entity,
+    mut player: Mut<AnimationPlayer>,
+    time: &Time,
+    animations: &Assets<AnimationClip>,
+    names: &Query<&Name>,
+    transforms: &mut Query<&mut Transform>,
+    children: &Query<&Children>,
+) {
+    let paused = player.paused;
+    // Continue if paused unless the `AnimationPlayer` was changed
+    // This allow the animation to still be updated if the player.elapsed field was manually updated in pause
+    if paused && !player.is_changed() {
+        return;
+    }
+
+    // Apply the main animation
+    _apply_animation(
+        1.0,
+        &mut player.animation,
+        paused,
+        entity,
+        time,
+        animations,
+        names,
+        transforms,
+        children,
+    );
+
+    // Apply any potential fade-out transitions from previous animations
+    for AnimationTransition {
+        current_weight,
+        animation,
+        ..
+    } in &mut player.transitions
+    {
+        _apply_animation(
+            *current_weight,
+            animation,
+            paused,
+            entity,
+            time,
+            animations,
+            names,
+            transforms,
+            children,
+        );
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn _apply_animation(
+    weight: f32,
+    animation: &mut PlayingAnimation,
+    paused: bool,
+    entity: Entity,
+    time: &Time,
+    animations: &Assets<AnimationClip>,
+    names: &Query<&Name>,
+    transforms: &mut Query<&mut Transform>,
+    children: &Query<&Children>,
+) {
+    if let Some(animation_clip) = animations.get(&animation.animation_clip) {
+        if !paused {
+            animation.elapsed += time.delta_seconds() * animation.speed;
+        }
+        let mut elapsed = animation.elapsed;
+        if animation.repeat {
+            elapsed %= animation_clip.duration;
+        }
+        if elapsed < 0.0 {
+            elapsed += animation_clip.duration;
```

Comment:
This is pre-PR code. GitHub doesn't show the changed lines very well. Not entirely sure why it is there but didn't want to touch it since there might be a corner case I missed.

Reply:
because animation can be played in reverse